File: python/max_subsequence.py


def max_subsequence(arr):
    start = 0
    stop = 0
    best = float("-inf")
    t = 0
    j = 0

    for i,x in enumerate(arr):
        t += x
        if t > best:
            best = t
            start = j
            stop = i
        elif t < 0:
            t = 0
            j = i+1

    return start, stop, best


if __name__ == "__main__":
    sample = [1, 2, -5, 4, 7, -2]
    start, stop, best = max_subsequence(sample)
    print(start, stop, best, sample[start:stop+1])

File: java/src/main/java/freezingoctosansa/MaxSubSequence.java
package freezingoctosansa;

// m[i] = max{a[i], m[i-1] + a[i]};
public class MaxSubSequence {

  public Integer Sum;
  public Integer Start;
  public Integer Stop;

  public void compute(Integer[] arr) {

    Start = 0;
    Stop = 0;
    Sum = Integer.MIN_VALUE;

    Integer t = 0;
    Integer j = 0;

    for (int i = 0; i < arr.length; i++) {
      t += arr[i];
      if (t > Sum) {
        Sum = t;
        Start = j;
        Stop = i+1;
      } if (t < 0) {
        t = 0;
        j = i+1;
      }
    }
  }
}
